src/key_exchange/tripledh.rs

```diff
@@ -387,44 +418,70 @@ fn derive_3dh_keys<KeyFormat: KeyPair, D: Hash>(
 
     let info: Vec<u8> = [
         STR_3DH,
-        &client_nonce,
-        &server_nonce,
-        &client_s_pk.to_arr(),
-        &server_s_pk.to_arr(),
+        &serialize(&client_nonce, 2),
+        &serialize(&server_nonce, 2),
+        &serialize(&client_s_pk.to_arr(), 2),
+        &serialize(&server_s_pk.to_arr(), 2),
     ]
     .concat();
 
-    const OUTPUT_SIZE: usize = 32;
-    let mut okm = [0u8; 3 * OUTPUT_SIZE];
-    let h = Hkdf::<D>::new(None, &ikm);
-    h.expand(&info, &mut okm)
-        .map_err(|_| InternalPakeError::HkdfError)?;
+    // TODO: could be made more efficient with re-using ikm
+    let handshake_secret = derive_secrets::<D>(&ikm, &STR_HANDSHAKE_SECRET, &info)?;
+    let session_secret = derive_secrets::<D>(&ikm, &STR_SESSION_SECRET, &info)?;
+    let km2 = hkdf_expand_label::<D>(
+        &handshake_secret,
+        &STR_SERVER_MAC,
+        b"",
+        <D as Digest>::OutputSize::to_usize(),
+    )?;
+    let km3 = hkdf_expand_label::<D>(
+        &handshake_secret,
+        &STR_CLIENT_MAC,
+        b"",
+        <D as Digest>::OutputSize::to_usize(),
+    )?;
+
     Ok((
-        GenericArray::clone_from_slice(&okm[..OUTPUT_SIZE]),
-        GenericArray::clone_from_slice(&okm[OUTPUT_SIZE..2 * OUTPUT_SIZE]),
-        GenericArray::clone_from_slice(&okm[2 * OUTPUT_SIZE..]),
+        GenericArray::clone_from_slice(&session_secret),
+        GenericArray::clone_from_slice(&km2),
+        GenericArray::clone_from_slice(&km3),
     ))
 }
 
-/// The third key exchange message
-pub struct KE3Message<HashLen: ArrayLength<u8>> {
-    mac: GenericArray<u8, HashLen>,
-}
+fn hkdf_expand_label<D: Hash>(
+    secret: &[u8],
+    label: &[u8],
+    context: &[u8],
+    length: usize,
+) -> Result<Vec<u8>, ProtocolError> {
+    let mut okm = vec![0u8; length];
 
-impl<HashLen: ArrayLength<u8>> ToBytes for KE3Message<HashLen> {
-    fn to_bytes(&self) -> Vec<u8> {
-        self.mac.to_vec()
-    }
-}
+    let mut hkdf_label: Vec<u8> = Vec::new();
+    hkdf_label.extend_from_slice(&length.to_be_bytes()[6..]);
 
-impl<HashLen: ArrayLength<u8>> TryFrom<&[u8]> for KE3Message<HashLen> {
-    type Error = InternalPakeError;
+    let mut opaque_label: Vec<u8> = Vec::new();
+    opaque_label.extend_from_slice(&STR_OPAQUE);
+    opaque_label.extend_from_slice(&label);
+    hkdf_label.extend_from_slice(&serialize(&opaque_label, 1));
```

Comment:
Do we want to hash in the length-prefixed serialized format, or simply the
bits themselves?
Edit: Answering my own question, [we want the serialization](https://github.com/cfrg/draft-irtf-cfrg-opaque/blob/master/draft-irtf-cfrg-opaque.md#3dh-key-derivation-derive-3dh).